apps/admin/frontend/src/components/print_diagnostic_button.tsx

```diff
@@ -0,0 +1,134 @@
+import { Button, Modal, P, RadioGroup } from '@votingworks/ui';
+import React, { useState } from 'react';
+import { assert, assertFalsy, sleep } from '@votingworks/basics';
+import type { DiagnosticsOutcome } from '@votingworks/admin-backend';
+import { addDiagnosticRecord, getPrinterStatus, printTestPage } from '../api';
+import { Loading } from './loading';
+
+/**
+ * Normally we want to free up the UI sooner, before the print is actually
+ * done, but in this case we want to wait until the print is done or almost
+ * done before presenting the verification step.
+ */
+export const TEST_PAGE_PRINT_DELAY_SECONDS = 7;
+
+type FlowState = 'printing' | 'verification' | 'test-failed';
+
+export function PrintTestPageButton(): JSX.Element {
+  const [flowState, setFlowState] = useState<FlowState>();
+  const [outcome, setOutcome] = useState<DiagnosticsOutcome>();
+
+  const printerStatusQuery = getPrinterStatus.useQuery();
+  const printTestPageMutation = printTestPage.useMutation();
+  const addDiagnosticRecordMutation = addDiagnosticRecord.useMutation();
+
+  const isPrinterConnected =
+    printerStatusQuery.isSuccess && printerStatusQuery.data.connected;
+
+  function resetFlow() {
+    setFlowState(undefined);
+    setOutcome(undefined);
+    printTestPageMutation.reset();
+    addDiagnosticRecordMutation.reset();
+  }
+
+  async function startFlow() {
+    printTestPageMutation.mutate();
+    setFlowState('printing');
+    await sleep(TEST_PAGE_PRINT_DELAY_SECONDS * 1000);
+    setFlowState('verification');
+  }
+
+  function verifyPrintOutcome() {
+    assert(outcome !== undefined);
+
+    addDiagnosticRecordMutation.mutate({
+      hardware: 'printer',
+      outcome,
+    });
```

Comment:
I thought it was fine in this case, because the test status line is going to update when the query updates (after the mutation succeeds anyway). The other side effects - closing the modal or showing the "Test Failed" modal, are fine to show.